### PR TITLE
Include the round number in block BLS signatures 

### LINKED ISSUE #595: BLS signatures need to include the round number
### Expected Behavior

Please describe the behavior you are expecting

### Current Behavior

What is the current behavior?

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR includes the round number in the data signed by validators. This allows double signing to be detected and slashed in the future.

### Tested

Unit and end-to-end tests.

### Other changes

- Refactor/renaming of the aggregated signature fields in IstanbulExtra

### Related issues

- Fixes #595 

### Backwards compatibility

Not backwards compatible

### CODE REVIEW

File: consensus/istanbul/backend/engine.go [hunk 1]
```diff
@@ -833,6 +839,30 @@ func ecrecover(header *types.Header) (common.Address, error) {
 	return addr, nil
 }
 
+func writeEmptyIstanbulExtra(header *types.Header) error {
+	log.Info("In write empty istanbul extra")
```
Comment (kevjue) [inline consensus/istanbul/backend/engine.go]: Trace?

File: consensus/istanbul/backend/engine.go [hunk 2]
```diff
@@ -379,6 +384,10 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
 		header.Time = big.NewInt(time.Now().Unix())
 	}
 
+	if err := writeEmptyIstanbulExtra(header); err != nil {
```
Comment (kevjue) [inline consensus/istanbul/backend/engine.go]: Is there a specific reason for this function?  From a quick glance, seems like it's putting zero values into all of the `IstanbulExtra` fields, which will happen by default when the header is constructed.

File: consensus/istanbul/backend/engine.go [hunk 3]
```diff
@@ -391,31 +400,28 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
...
-		bitmap := parentExtra.Bitmap
-		asig := parentExtra.CommittedSeal
+		parentAggregatedSeal := parentExtra.AggregatedSeal
 
-		parentGossipedSeals := sb.core.ParentCommits()
-		if parentGossipedSeals != nil && parentGossipedSeals.Size() != 0 {
-			sb.logger.Trace("backend.Prepare: taking union of parent seals", "valset", parentGossipedSeals.String(), "num", number)
+		additionalParentSeals := sb.core.ParentCommits()
+		if additionalParentSeals != nil && additionalParentSeals.Size() != 0 {
+			logger.Trace("Combining additional seals with the parent aggregated seal", "valset", additionalParentSeals.String(), "num", number, "parentAggregatedSeal", parentAggregatedSeal.String())
```
Comment (kevjue) [inline consensus/istanbul/backend/engine.go]: nit:  Should `valset` be renamed to `additionalParentSeals`?

File: consensus/istanbul/backend/engine.go [hunk 4]
```diff
@@ -391,31 +400,28 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
...
-			if err := sb.checkValidatorSignatures(parent.Hash(), parentValidators, newBitmap, newAsig); err != nil {
-				sb.logger.Error("backend.Prepare: tried to create invalid aggregate signature. not updating to union")
+			if err := sb.verifyAggregatedSeal(parent.Hash(), parentValidators, unionAggregatedSeal); err != nil {
+				logger.Error("Failed to combine additional seals with parent aggregated seal.")
 			} else {
-				bitmap = newBitmap
-				asig = newAsig
-				sb.logger.Debug("backend.Prepare: parent seal updated!", "bitmap", bitmap.Bits(), "sig", common.ToHex(asig))
+				parentAggregatedSeal = unionAggregatedSeal
+				logger.Debug("Succeeded in combine additional seals with parent aggregated seal", "combinedAggregatedSeal", parentAggregatedSeal.String())
```
Comment (kevjue) [inline consensus/istanbul/backend/engine.go]: nit:  `Succeeded in combining ...`

File: core/types/istanbul_test.go [hunk 5]
```diff
@@ -63,12 +64,9 @@ func TestExtractToIstanbul(t *testing.T) {
 				},
 				AddedValidatorsPublicKeys: [][]byte{},
 				RemovedValidators:         big.NewInt(12), //1100
-				Bitmap:                    big.NewInt(0),
-				Seal:                      []byte{},
```
Comment (kevjue) [inline core/types/istanbul_test.go]: nit:  Was removal of the `Seal` field unintentional?

File: consensus/istanbul/types.go [hunk 6]
```diff
@@ -31,6 +31,8 @@ type Proposal interface {
 	// Number retrieves the sequence number of this proposal.
 	Number() *big.Int
 
+	Header() *types.Header
```
Comment (kevjue) [inline consensus/istanbul/types.go]: I may have missed this, but what's the purpose of this new field?  I don't recall seeing it used anywhere in this PR.

File: consensus/istanbul/backend/engine.go (hunk 1 above)
Comment (asaj) [reply consensus/istanbul/backend/engine.go]: Whoops! Should have been removed

File: consensus/istanbul/backend/engine.go (hunk 3 above)
Comment (asaj) [reply consensus/istanbul/backend/engine.go]: IMO `additionalParentSeals` is a more readable name

File: core/types/istanbul_test.go (hunk 5 above)
Comment (asaj) [reply core/types/istanbul_test.go]: Nope!

File: consensus/istanbul/core/core.go [hunk 7]
```diff
@@ -186,41 +190,37 @@ func (c *core) commit() {
 // AggregateSeals aggregates all the given seals for a given message set to a bls aggregated
 // signature and bitmap
 // TODO: Maybe return an error instead of panicking?
```
Comment (gakonst) [inline consensus/istanbul/core/core.go]: @asaj this todo can be removed now

**Pull request merged**